config/impl/src/main/java/org/jboss/windup/config/loader/WindupRuleLoaderImpl.java

```diff
@@ -201,4 +208,61 @@ private String generatedRuleID(WindupRuleProvider provider, Rule rule, int idx)
         String provID = provider.getID().replace("org.jboss.windup.rules.", "w:");
         return "GeneratedID_" + provID + "_" + idx;
     }
+
+
+    /**
+     * Processes metadata - collects them from various sources (API alternatives),
+     * validates and normalizes them.
+     */
+    private void processRuleMetadata(Rule rule, WindupRuleProvider provider)
+    {
+        final Context ctx = (Context) rule;
+
+        provider.enhanceMetadata(ctx);
+
+        normalizeRuleMetadata(rule);
+        assert ctx.get(RuleMetadata.CATEGORY) instanceof Set;
+
+        Set<String> categories = (Set<String>) ctx.get(RuleMetadata.CATEGORY);
+        categories.addAll(provider.getCategories());
+    }
+
+
+    /**
+     * Converts metadata to a normalized form (i.e. coma-separated String to Collection etc.
+     */
+    private static void normalizeRuleMetadata(Rule rule)
+    {
+        if (!(rule instanceof Context))
+            return;
+
+        Context ctx = (Context) rule;
```

Comment:
ctx is very common, almost standard abbrev for context, used accross JBoss, even within Windup.